DESIGN.md

```diff
@@ -0,0 +1,237 @@
+    there.
+        - First command run must be `config`. Will behave similarly to the AWS CLI `config` by setting some default
+        connection and credential info which will be used in subsequent commmands.
+        - Commands will include `start`, `stop`, `status`, `output`, `take`, and `list` just like the HTTP API.
+        These commands will fail if the CLI has not been configured.
+    - *worker*: The actual Linux service to manage jobs. Using a service is convenient because it provides 
+    automatic restarts if the service crashes due to somebody running something that consumes enormous RAM or 
+    something.
+        - Exposes the methods for Start, Stop, Status, Output, List, and Take via RPC
+- *Makefile*: will build one or all of the 3 executables
+
+
+### Worker
+The job worker will make use of `net/rpc` for communicating with the HTTP server. The package is in feature lock, but 
+provides an easy interface for IPC and development is still current. There have been new commits to the package that
+will be included in the Go 1.17 release. Using Unix domain sockets will allow pretty low latency and is supported by 
+net/rpc. Better performance could be possible, but won't be necessary for this example. 
+
+The `os/exec` package allows capturing stdout/stderr directly and store it in the job. Not memory efficient to keep it 
+all in RAM, but quick to implement. Storing stdout and stderr will be done separately to allow the user to get get 
+them as two separate fields when retreiving output.
+
+Worker will initialize logging, initialize worker hub, and intialize an RPC listener. Worker hub will contain a list
+of all workers and a go routine to occasionally prune workers that have stopped. A worker will be considered stale 
+after it has been stopped for 1 hour and its output will be discarded.
+
+Job type will include the job owner id, the command that started it, times for different events such as start 
+and finish, byte slices of stderr and stdout and other relevant information required by the API and CLI. Listener
+will make a few methods available via RPC for use by the API and CLI clients. The List method will only list jobs
+owned by the users passed in as an argument, or all if there are no users passed in.
+
+Each job will be protected by a mutex to prevent conflicting commands.
+
+The hub will have a map of jobID to jobs. The map will also be protected by a mutex to prevent simultaneous access.
+
+### API
+Will use the `net/http` package for ease of use and fast development. Server will be configured for mTLS with TLS
+version 1.3 only and a small set of ciphers. The HTTP server will verify the client TLS certificates and 
+middleware will perform the authorization before passing requests off to the handlers. As there are only 6 
+endpoints, no third-party mux will be used and path variables will be parsed manually.
```

Comment:
Could you elaborate based on what the server will validate client certificate ?

Reply:
I'll generate a Certificate Authority cert which will be used to sign both the server and client certificates. The CA cert will also be added to the list of trusted CAs on both the client and server. Therefore the server will be verifying that the client has supplied a valid certificate signed by the CA cert.